pow/src/main/java/tech/pegasys/artemis/pow/Eth1Manager.java

```diff
@@ -13,113 +13,121 @@
 
 package tech.pegasys.artemis.pow;
 
-import static tech.pegasys.artemis.util.config.Constants.ETH1_FOLLOW_DISTANCE;
+import static com.google.common.base.Preconditions.checkNotNull;
 
 import com.google.common.primitives.UnsignedLong;
-import io.reactivex.disposables.Disposable;
 import java.math.BigInteger;
-import java.util.concurrent.TimeUnit;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 import org.apache.tuweni.bytes.Bytes32;
 import org.web3j.protocol.core.methods.response.EthBlock;
-import tech.pegasys.artemis.pow.api.MinGenesisTimeBlockEventChannel;
+import tech.pegasys.artemis.pow.api.Eth1EventsChannel;
 import tech.pegasys.artemis.pow.event.MinGenesisTimeBlockEvent;
 import tech.pegasys.artemis.util.async.AsyncRunner;
 import tech.pegasys.artemis.util.async.SafeFuture;
 import tech.pegasys.artemis.util.config.Constants;
 
-public class Eth1MinGenesisTimeBlockFinder {
+public class Eth1Manager {
 
   private static final Logger LOG = LogManager.getLogger();
 
   private final Eth1Provider eth1Provider;
-  private final MinGenesisTimeBlockEventChannel minGenesisTimeBlockEventChannel;
   private final AsyncRunner asyncRunner;
+  private final Eth1EventsChannel eth1EventsChannel;
+  private final DepositProcessingController depositProcessingController;
 
-  private volatile Disposable latestBlockDisposable;
-
-  public Eth1MinGenesisTimeBlockFinder(
+  public Eth1Manager(
       Eth1Provider eth1Provider,
-      MinGenesisTimeBlockEventChannel minGenesisTimeBlockEventChannel,
-      AsyncRunner asyncRunner) {
+      AsyncRunner asyncRunner,
+      Eth1EventsChannel eth1EventsChannel,
+      DepositProcessingController depositProcessingController) {
     this.eth1Provider = eth1Provider;
-    this.minGenesisTimeBlockEventChannel = minGenesisTimeBlockEventChannel;
     this.asyncRunner = asyncRunner;
+    this.eth1EventsChannel = eth1EventsChannel;
+    this.depositProcessingController = depositProcessingController;
   }
 
   public void start() {
-    findAndPublishFirstValidBlock()
-        .finish(
-            () ->
-                LOG.info(
-                    "Eth1MinGenesisBlockFinder successfully found first "
-                        + "(time) valid genesis block"));
+
+    SafeFuture<EthBlock.Block> headBlock = getHead();
+
+    SafeFuture<Boolean> isHeadBeforeMinGenesis = isHeadBeforeMinGenesis(headBlock);
+
+    isHeadBeforeMinGenesis.thenAccept(
+        bool -> {
+          if (bool) headBeforeMinGenesisMode(headBlock);
+          else headAfterMinGenesisMode(headBlock);
+        });
```

Comment:
The future chaining here is making life much more complex than it needs to be.  First, get the head, then when you have the head, decide if it's before min genesis and call the right method.  Right now you're starting the process to get head, then queuing up the process to decide if it's before or after, then queuing up the decision of which method to call etc.

It should just be:
```
  public void start() {
    getHead().thenCompose(headBlock -> {
      if (isHeadBeforeMinGenesis(headBlock)) {
        return headBeforeMinGenesisMode(headBlock);
      } else {
        return headAfterMinGenesisMode(headBlock);
      }
    });
  }
```

and `isHeadBeforeMinGenesis`, `headBeforeMinGenesisMode` and `headAfterMinGenesisMode` all take a `EthBlock.Block` rather than a `SafeFuture<EthBlock.Block>`.  Now life is easy. :)

The other detail is that you need to handle errors - mostly if `getHead` fails or the requests for deposits in `headBeforeMinGenesisMode` and `headAfterMinGenesisMode` fails.

Reply:
Makes perfect sense. Done.

Reply:
I'm now dealing with the error case of getHead in the getHead function itself, and I plan to do the same for headBeforeMinGenesis etc as well because it gives me more granular control. Does that make sense?